File: book/continuous-time/java/Signal.java
import nz.sodium.*;
import java.util.Optional;

public class Signal {
    public Signal(double t0, double a, double b, double c) {
        this.t0 = t0;
        this.a = a;
        this.b = b;
        this.c = c;
    }
    public final double t0, a, b, c;
    public double valueAt(double t) {
        double x = t - t0;
        return a*x*x + b*x + c;
    }
    public final static double quantum = 0.000001;
    public Optional<Double> when(double x) {
        double c = this.c - x;
        if (a == 0) {
            double t = (-c) / b;
            return t >= quantum ? Optional.of(t + t0)
                                : Optional.empty();
        }
        else {
            double b24ac = Math.sqrt(b*b - 4*a*c);
            double t1 = ((-b) + b24ac) / (2*a);
            double t2 = ((-b) - b24ac) / (2*a);
            return t1 >= quantum
                ? t2 >= quantum ? Optional.of((t1 < t2 ? t1 : t2) + t0)
                                : Optional.of(t1 + t0)
                : t2 >= quantum ? Optional.of(t2 + t0)
                                : Optional.empty();
        }
    }
    public Signal integrate(double initial) {
        if (a != 0.0) throw new InternalError("Signal can't handle x^3");
        return new Signal(t0, b/2, c, initial);
    }
    public static Cell<Signal> integrate(
                                    Cell<Signal> sig, double initial) {
        Stream<Signal> sSig = Operational.updates(sig);
        return sSig.accum(sig.sample().integrate(initial),
            (neu, old) -> neu.integrate(old.valueAt(neu.t0)));
    }
}

File: book/continuous-time/c#/Shared/Signal.cs
using System;
using Sodium.Frp;
using Sodium.Functional;

namespace Shared
{
    public class Signal
    {
        public static readonly double Quantum = 0.000001;

        public Signal(double t0, double a, double b, double c)
        {
            this.T0 = t0;
            this.A = a;
            this.B = b;
            this.C = c;
        }

        public double T0 { get; }
        public double A { get; }
        public double B { get; }
        public double C { get; }

        private static bool IsCloseTo(double x, double y)
        {
            return Math.Abs(x - y) < Quantum;
        }

        public double ValueAt(double t)
        {
            double x = t - this.T0;
            return this.A * x * x + this.B * x + this.C;
        }

        public Maybe<double> When(double x)
        {
            double c = this.C - x;
            if (IsCloseTo(this.A, 0))
            {
                double t = -c / this.B;
                return t >= Quantum ? Maybe.Some(t + this.T0) : Maybe.None;
            }

            // ReSharper disable once InconsistentNaming
            double b24ac = Math.Sqrt(this.B * this.B - 4 * this.A * c);
            double t1 = (-this.B + b24ac) / (2 * this.A);
            double t2 = (-this.B - b24ac) / (2 * this.A);
            return t1 >= Quantum
                ? t2 >= Quantum
                    ? Maybe.Some((t1 < t2 ? t1 : t2) + this.T0)
                    : Maybe.Some(t1 + this.T0)
                : t2 >= Quantum
                    ? Maybe.Some(t2 + this.T0)
                    : Maybe.None;
        }

        public Signal Integrate(double initial)
        {
            if (!IsCloseTo(this.A, 0.0))
            {
                throw new Exception("Signal can't handle x^3");
            }

            return new Signal(this.T0, this.B / 2, this.C, initial);
        }

        public static Cell<Signal> Integrate(Cell<Signal> sig, double initial)
        {
            Stream<Signal> sSig = sig.Updates();
            return sSig.Accum(sig.Sample().Integrate(initial), (n, o) => n.Integrate(o.ValueAt(n.T0)));
        }
    }
}